### PR TITLE
print out dependency build failure during cabal build / repl

### LINKED ISSUE #5974: `new-build` returns exit code 1 (but doesn't indicate that it fails) when run in a cabal project
Consider this `.cabal` file:

```cabal
cabal-version:       >=1.10
name:                cabal-sandbox
version:             0.1.0.0
build-type:          Simple

library
  exposed-modules:     CabalSandbox
  build-depends:       base, FPretty
  hs-source-dirs:      src
  default-language:    Haskell2010
```

I'm choosing to depend on `FPretty` here since it's a package that's known not to build correctly against GHC 8.4 or later. If you try building `FPretty` in the same directory as `cabal-sandbox.cabal` using `new-build`, then this happens:

```
$ /opt/cabal/3.0/bin/cabal new-build FPretty --allow-newer=base -w /opt/ghc/8.6.4/bin/ghc
Resolving dependencies...
Build profile: -w ghc-8.6.4 -O1
In order, the following will be built (use -v for more details):
 - FPretty-1.1 (lib) (requires build)
Starting     FPretty-1.1 (lib)
Building     FPretty-1.1 (lib)
```

That's all the output that appears. One might think that this means that `FPretty-1.1` was built successfully, but it turns out that that's not the case at all. In fact, this command actually _fails_, which is something that you can confirm if you check the exit code:

```
$ echo $?
1
```

Why, then, does an error message not appear? Quite strange.

This happens with other commands as well, such as `new-repl`:

```
$ /opt/cabal/3.0/bin/cabal new-repl FPretty --allow-newer=base -w /opt/ghc/8.6.4/bin/ghc
Build profile: -w ghc-8.6.4 -O1
In order, the following will be built (use -v for more details):
 - FPretty-1.1 (lib) (requires build)
Starting     FPretty-1.1 (lib)
Building     FPretty-1.1 (lib)

$ echo $?
1
```

### PR DESCRIPTION
**Pull request by ulysses4ever**
fix #5974 #7727

This is a cheap fix via straightforward extension of #6102

---
Please include the following checklist in your PR:

* [ ] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#conventions).
* [ ] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [ ] The documentation has been updated, if necessary.

Please also shortly describe how you tested your change. Bonus points for added tests!

### CODE REVIEW

**Comment by ulysses4ever** [comment]
That’s better: we pass CI now. Next, need a test showing actual profit from it as the current tests that I had to update are not very convincing unfortunately.

**Comment by Mikolaj** [comment]
Does the runner run with -j2? Does it apply any special segregation of the outputs?

**Comment by jneira** [comment]
just in case it could help, the test runner uses 2 jobs: https://github.com/haskell/cabal/blob/12094335096ff887804828bf638a9d76b5daf25f/.github/workflows/validate.yml#L30

**Comment by ulysses4ever** [comment]
@jneira ooo... I knew there should be something like this! One thing, though: I don't see an effect of `-j` in test scripts even when running them manually, without `validate.sh`. Do you think this line explains that? 

(I should perhaps upload the test I'm trying, so that you see it even if it doesn't work at this point.)

**Comment by ulysses4ever** [comment]
@Mikolaj I want to petition for this to be merged without a test:

1) It does fix #5974 #7727 locally for me.

2) Update in the existing tests (see [Files changed](https://github.com/haskell/cabal/pull/8296/files)) shows that it makes output for `-j1` more instructive.

3) (Most importantly) The test suite insists on `-j1` (judging by the commit message, in the interest of determinism in test outputs):

   https://github.com/haskell/cabal/blob/410f871df899e5af0847089354e0031fe051551d/cabal-testsuite/src/Test/Cabal/Prelude.hs#L305-L308

    Therefore, any change that concerns `-j2` (or more than 2), which this one is an instance of, can't be tested.

**Comment by Mikolaj** [comment]
I see. So the runner runs with -j2, but inside it, (most of) out tests run with -j1?

I trust your judgement on that. Let's skip adding more tests.

File: cabal-install/src/Distribution/Client/ScriptUtils.hs [hunk 1]
```diff
@@ -217,11 +218,11 @@ withContextAndSelectors noTargets kind flags@NixStyleFlags {..} targetStrings gl
     defaultTarget = [TargetPackage TargetExplicitNamed [fakePackageId] Nothing]
 
     with = do
-      ctx <- establishProjectBaseContext verbosity cliConfig OtherCommand
+      ctx <- establishProjectBaseContext verbosity cliConfig cmd
```
Comment (Mikolaj) [inline cabal-install/src/Distribution/Client/ScriptUtils.hs:221]: And also a good catch.

**Review by Mikolaj — APPROVED** [review]
LGTM.

**Comment by ulysses4ever** [comment]
> I see. So the runner runs with -j2, but inside it, (most of) out tests run with -j1?

In a nutshell, yes. That j1 is what is passed to the cabal being tested, whereas validate’s j2 is what gets the cabal that executes the test suite.

**Comment by ulysses4ever** [comment]
@jneira by any chance, could you take another look here?

**Review by jneira — APPROVED** [review]
lgtm, not only the ux is improved but the types are more precise so i hope future improvements will be easier
small note: iiuc there is no test over cabal repl and no previous one had to be updated, have you checked them?

**Comment by ulysses4ever** [comment]
@jneira thank you for the review. Could you, perhaps, expand a little bit on what you think could be checked about repl? I just checked that cabal from this PR gives the same error for `cabal repl` as it does for `cabal build` under conditions described in #5974. Which makes sense to me.

I think important thing to understand about this PR is that it changes the behavior observable only in the case of an unrecoverable error during the build. So, hopefully, it's hard to screw up things under this scenario: the user will only see any difference if they are not getting anything useful out of a build anyway. So, hopefully, it's a low-stake change.

**Comment by jneira** [comment]
I observed repl command was included for show the new message and I did not see a repl test changed to take the new msg in account, but maybe it should not?

**Comment by ulysses4ever** [comment]
@jneira indeed, and thanks for noticing (I completely forgot about Repl by now). The reason that no repl tests have to be updated is simply because there seems to be no `fails` tests for repl. Do you want me to add one?

**Comment by jneira** [comment]
well not as requirement to merge I'd you have tested it manually but the fails test could help in ongoing PRS updating it, the same way the previous ones helped this pr

**Comment by ulysses4ever** [comment]
I was playing with `cabal repl` in the test suite for a bit, and I don't see how to go about it. It doesn't seem like `cabal repl` actually hits `dieIfNotHaddockFailure` (changed in this patch). At least, not for the kind of errors I tried.

I tried erroring local package, and erroring in a dependency supplied by `withRepo`. The latter is closer to what this patch was trying to address, so here's even more detail. The test:
```
main = cabalTest . withRepo "repo" . void $ do
    cabalWithStdin "v2-repl" ["--build-depends", "buggy-pkg"] ":q"
```

No matter what I tried, this patch doesn't seem to affect `cabal repl` output; or, more precisely, the bits recorded by the test suite doesn't seem to be affected. 

In particular, for `v2-build` the failing output (that I did have to update with this patch) is:
```
# cabal v2-build
Resolving dependencies...
Build profile: -w ghc-<GHCVER> -O1
In order, the following will be built:
 - example-1.0 (lib) (first run)
Configuring library for example-1.0..
Preprocessing library for example-1.0..
Building library for example-1.0..
Error: cabal: Failed to build example-1.0-inplace.
```
The last line was added by this PR. But `v2-repl` shows:
```
# cabal v2-repl
Resolving dependencies...
Build profile: -w ghc-<GHCVER> -O1
In order, the following will be built:
 - example-1.0 (lib) (first run)
Configuring library for example-1.0..
Preprocessing library for example-1.0..
```
And that's it... (again, this is `cabal.out` file that the test suite creates).

---

I really don't want to spend much more time on this. Please, scream if you have ideas how to improve `repl` testing. Alternatively, I could erase mentions of `ReplCommand` in this patch (it will still address the issues I reference). I think it'd a downturn because I think no matter the tests it's better to track commands we're running rather then not doing that. But I can see how a non-testable thing feels repelling.

**Pull request merged**